{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sloppiness, robustness and evolvability of discrete and continuous biochemical models\n",
    "\n",
    "Models in systems biology aim to describe key processes in\n",
    "single and population of cells which can be represented by complex\n",
    "biochemical networks. Depending on the availability of\n",
    "experimental data, these biochemical networks can be\n",
    "modeled as either a set of ordinary differential equations\n",
    "(deterministic/stochastic) which are continuous in time \n",
    "or as a discrete set of\n",
    "difference equations where time is discrete and each\n",
    "species can only assume a high or low value. These models \n",
    "are approximate and often are poorly parametrized\n",
    "because the necessary rate constants have not been measured\n",
    "and/or the absolute concentration of cellular biochemical species\n",
    "are not typically known. Also these networks are almost always\n",
    "incompletely known as many components and interactions that play\n",
    "a role in determining the final outcome are yet to be\n",
    "identified. Because of this, the predictability of these models\n",
    "are poor in general. Even when we have a model that can\n",
    "explain the experimental data for particular situations \n",
    "it is not sufficient to answer some of\n",
    "the key questions biologists are interested in. These have\n",
    "to do with the concepts of robustness and evolvability.\n",
    "For example:\n",
    "\n",
    "- If the network is perturbed due to mutations or\n",
    "  internal fluctuations, is it robust enough to maintain\n",
    "  its functions? \n",
    "- If there is a change in the environmental conditions or \n",
    "  if the system is subjected to external stress/stimuli, \n",
    "  can the system adapt itself to cope with the changes? \n",
    "\n",
    "In this review we will define these concepts more carefully, \n",
    "give some mathamatical interpretations and apply them to \n",
    "some simple scenarios."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3><a id=\"toc\"> Table of Contents</a></h3>\n",
    "\n",
    "- [Qualtitative Discussion](#Qual_Disc)\n",
    "    * [Sloppiness](#Sloppiness_Qual)\n",
    "    * [Robustness](#Robustness_Qual)\n",
    "    * [Evolvability](#Evolvability_Qual)\n",
    "- [Detailed Description and Mathematical Formulation](#Math_Detail)\n",
    "    * [Kinetic ensemble method and sloppiness](#kinetic_ensemble)\n",
    "    * [Robustness and Evolvability](#detail_robust_evolve)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3><a id='Qual_Disc'>Qualitative Discussion</a></h3>[[top](#toc)]\n",
    "\n",
    "<h4><a id='Sloppiness_Qual'>Sloppiness:</a></h4> As mentioned in the introduction the systems\n",
    "biology models are ill-conditioned and the collective\n",
    "behavior cannot be in general used to infer the underlying\n",
    "parameters [(Daniels, 2008)](#daniels2008sloppiness). However most\n",
    "important models in systems biology have been shown to be\n",
    "*sloppy* [(Gutenkunst, 2007)](#gutenkunst2007universally) in the sense that \n",
    "the feasible parameter space of the model is highly anisotropic \n",
    "with **few stiff and many soft directions**. What this means is that\n",
    "useful model predictions can be made even when a large set of\n",
    "parameters are unknown or known with high degree of uncertainty. \n",
    "However in face of such parameter\n",
    "uncertainties an attempt to find a single best model that\n",
    "explains the experimental observation is a futile endeavor.\n",
    "What will be more useful is to obtain an ensemble of model\n",
    "which are constrained by experimental data. Such an ensemble\n",
    "makes a model more falsifiable because any new experimental\n",
    "measurement can be compared with the ensemble deviation.\n",
    "The model will be considered defective if new data falls\n",
    "outside this deviation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": false
   },
   "source": [
    "<figure>\n",
    "<img src=\"Sethna_Universal_Sloppiness.png\" alt=\"Sloppiness in Biological Mdoels\" style=\"width: 500px;\"/>\n",
    "<figcaption>Sloppiness in Biological Models</figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4><a id='Robustness_Qual'>Robustness:</a></h4> There has been much debate over the correct\n",
    "definition of robustness in biology and a true definition\n",
    "has not yet been found. But most agree with the basic\n",
    "notion of robustness. We describe the idea of robustness\n",
    "following the pioneering work of Kitano  [(Kitano, 2007)](#kitano2007towards)\n",
    "and Wagner [(Ciliberti, 2007)](#ciliberti2007robustness). While describing\n",
    "robustness we must keep in mind that it is distinct from\n",
    "related concepts of **homeostasis** and **stability**. The key\n",
    "difference is that in the face of perturbations and\n",
    "uncertainties, homeostasis or stability tries to\n",
    "maintain the **state** of the system whereas robustness tries\n",
    "to maintain **function**. Thus a robust system can (and actually does)\n",
    "move itself to a different stable state or promote\n",
    "instability in the face of external stress or change in\n",
    "environmental conditions to maintain system functions.\n",
    "Fo example, certain type of bacteria undergoes a\n",
    "phenotypic switch in the face of harsh external conditions.\n",
    "Cancer cells in tumor survive chemotherapy by promoting\n",
    "instability - giving rise to a heterogeneous population.\n",
    "The key questions regarding robustness of a biological\n",
    "system are:\n",
    "\n",
    "- Is it conserved?\n",
    "- What trade-offs exist in biological system between robustness and fragility?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "<figure>\n",
    "<img src=\"Kitano_Robustness.jpg\" style=\"width: 500px;\"/>\n",
    "<figcaption>Robustness in biological system  [(Kitano, 2007)](#kitano2007towards)</figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4><a id='Evolvability_Qual'>Evolvability:</a></h4> Evolvability is even harder to define\n",
    "and is very much dependent on particular biological context\n",
    "and time scales  [(Masel, 2010)](#masel2010robustness). At the longest geological time scales\n",
    "we have evolvability of major morphological features and\n",
    "other radical innovations whereas at the shortest time\n",
    "scale we can consider single generation evolvability\n",
    "captured entirely by quantitative genetics. For systems\n",
    "biology model we are at an intermediate scale governed\n",
    "by population genetics where we want to observe the\n",
    "capacity of the system to produce phenotypes that are\n",
    "not unconditionally deleterious or fails to maintain\n",
    "essential function for survival."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TABLE 1: The 'conceptual spectrum' of evolvability [Pigliucci, 2008](#pigliucci2008)\n",
    "\n",
    "<table><colgroup><col align=\"left\"/><col align=\"left\"/><col align=\"left\"/><col align=\"left\"/><col align=\"left\"/></colgroup><thead valign=\"top\"><tr><th class=\"last rowsep\">Suggested term</th><th class=\"last rowsep\">Scale</th><th class=\"last rowsep\">Description</th><th class=\"last rowsep\">Effects</th><th class=\"last rowsep\">Example</th></tr></thead><tbody valign=\"top\"><tr class=\"odd\"><td>Heritability (<span class=\"i\">sensu</span> Houle)</td><td>Within populations</td><td>Standing pool of genetic variation and covariation</td><td>Determines the response to natural selection within populations</td><td class=\"image\"><img alt=\"Is evolvability evolvable?\" src=\"nrg2278-i1.jpg\"/></td></tr><tr><td>Evolvability (<span class=\"i\">sensu</span> Wagner &amp; Altenberg)</td><td>Within species</td><td>Includes variability (<span class=\"i\">sensu</span> Wagner &amp; Altenberg), depends on genetic architecture and developmental constraints</td><td>Affects long-term adaptation, channels evolution along non-random trajectories, allows mid-term exploration of phenotypic space</td><td class=\"image\"><img alt=\"Is evolvability evolvable?\" src=\"nrg2278-i2.jpg\"/></td></tr><tr class=\"odd\"><td class=\"last\">Innovation (<span class=\"i\">sensu</span> Maynard-Smith &amp; Szathmary)</td><td class=\"last\">Within clades</td><td class=\"last\">As for within species, but includes the capacity to overcome standing genetic and developmental constraints, opening new areas of phenotypic space for further evolution</td><td class=\"last\">Generates major phenotypic (morphological, behavioural or physiological) breakthroughs (novelties)</td><td class=\"last image\"><img alt=\"Is evolvability evolvable?\" src=\"nrg2278-i3.jpg\"/></td></tr></tbody></table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3><a id=\"Math_Detail\">Detailed Description and Mathematical Formulation</a></h3>[[top](#toc)]\n",
    "\n",
    "<h4><a id=\"kinetic_ensemble\"> Kinetic ensemble method and sloppiness:</a></h4>\n",
    "\n",
    "The basic principle of the ensemble method [(Battogtokh, 2002)](#battogtokh2002ensemble)\n",
    "is to generate an experimentally constrained random\n",
    "sample of the vector $\\bar{\\Theta}$, which is the set of all parameters\n",
    "and species concentrations that specifies a model. The degree of consistency of\n",
    "model with the experimental data can be quantified by\n",
    "calculating a certain figure of merit or a cost function.\n",
    "Let $\\bar{Y} := (Y_1, ..., Y_D)$ is the D-tuple of all experimental\n",
    "observables. $D = M_S \\times M_T \\times M_E$ where:\n",
    "\n",
    "- $M_E$ different realization of an experiment is performed.\n",
    "  Each realization is distinguished by various conditions\n",
    "  like external growth factor concentrations.\n",
    "- $M_S$ distinct species are present whose concentrations\n",
    "  are being measured as a function of time\n",
    "- Concentrations are recorded for $M_T$ different time points"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let $F(\\bar{\\Theta})$ be the vector of values corresponding\n",
    "to Y predicted from the model. Considering the experimental data\n",
    "is distributed normally (not essential) the probability\n",
    "distribution of the vector of observables $\\bar{Y}$ given the\n",
    "mean $\\bar{\\mu}$\n",
    "\n",
    "$$P(\\bar{Y}|\\bar{\\mu}) \\propto exp[-\\sum_{l}\\frac{(Y_i - \\mu_i)^2}{2\\sigma_l^2}]$$\n",
    "\n",
    "Our quantity of interest here the probability distribution\n",
    "$Q(\\bar{\\Theta})$ of the parameters. Considering the experimental\n",
    "data we want to calculate the conditional probability\n",
    "$Q(\\bar{\\Theta}|\\bar{Y})$ which we can write using a Bayesian interpretation\n",
    "as\n",
    "\n",
    "$$Q(\\bar{\\Theta}|\\bar{Y}) = \\frac{p(\\bar{\\Theta})P(\\bar{Y}|F(\\bar{\\Theta}))}{\\int p(\\bar{\\Theta})P(\\bar{Y}|F(\\bar{\\Theta}))}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we choose a flat prior the above equation can be\n",
    "modified as\n",
    "\n",
    "$$Q(\\bar{\\Theta}|\\bar{Y}) = \\frac{P(\\bar{Y}|F(\\bar{\\Theta}))}{\\int P(\\bar{Y}|F(\\bar{\\Theta}))} = \\frac{P(\\bar{Y}|F(\\bar{\\Theta}))}{\\Omega}$$\n",
    "\n",
    "The probability $P(\\bar{Y}|F(\\bar{\\Theta}))$ can be written as\n",
    "\n",
    "$$P(\\bar{Y}|F(\\bar{\\Theta})) \\propto exp[-\\sum_{l=1}^{D}\\frac{(Y_i - F_i(\\bar{\\Theta}))^2}{2\\sigma_l^2}]$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sethna et. al. [Gutenkunst, 2007](#gutenkunst2007universally) defines\n",
    "the factor inside the exponential as the cost function \n",
    "with a slighly modified expression as below\n",
    "\n",
    "$$C(\\bar{\\Theta}) = [\\sum_{l=1}^{D}\\frac{(Y_i - B_kF_i(\\bar{\\Theta}))^2}{2\\sigma_l^2}] + f(F(\\bar{\\Theta}))$$\n",
    "\n",
    "Here the factors $B_k$ consists of undetermined multiplicative\n",
    "constants that restricts the interpretation of experimental\n",
    "data and nonlinear function f allows us to enter some fuzzy data\n",
    "like inequalities.\n",
    "\n",
    "Hence we can write the expression for conditional probability\n",
    "$Q(\\Theta|Y)$ as proportional to \n",
    "\n",
    "$$Q(\\Theta|Y) \\propto exp[-\\beta C(\\bar{\\Theta})]$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hence using this ensemble we can setup a Monte Carlo scheme\n",
    "and perform moves in the\n",
    "parameter space $\\bar{\\Theta}$ and accept those that decreases\n",
    "the cost function and accept with a probability those\n",
    "that increase the cost function. Hence for any observable\n",
    "$O$ like a time dependent chemical concentration we can\n",
    "calculate the average and the standard deviation as below\n",
    "\n",
    "$$<O> = \\frac{1}{N_E}\\sum_{j=1}^{D}O_j$$\n",
    "$$\\sigma_O = \\sqrt{<O^2> - <O>^2}$$\n",
    "\n",
    "From the definition of cost functions we can identify\n",
    "the stiff and soft directions as below by computing the\n",
    "Hessian matrix defined below\n",
    "\n",
    "$$H_{ij}(\\theta^*) = \\frac{\\partial^2C}{\\partial\\theta_i\\partial\\theta_j}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively in the spirit of the ensemble method one can\n",
    "construct a covariance matrix using the ensemble data as\n",
    "below.\n",
    "\n",
    "$$\\Theta = <(\\theta - <\\theta>)(\\theta - <\\theta>)^T>$$\n",
    "\n",
    "An eigenvalue decomposition of the matrix will also give\n",
    "us the stiff and soft directions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4><a id=\"detail_robust_evolve\">Robustness and Evolvability</a></h4>[[top](#toc)]\n",
    "\n",
    "<h5><a id=\"detail_robust_sub\">Robustness</a></h5>\n",
    "\n",
    "Kitano [(Kitano, 2007)](#kitano2007towards) defines robustness taking\n",
    "into account his definition - which is the ability of a\n",
    "system to maintain its functions in face of perturbations.\n",
    "So the robustness $R_{aP}^s$ of a system s with regard to\n",
    "its function a can be defined as\n",
    "\n",
    "$$R_{aP}^s = \\int_{P}\\psi(P)D_{a,p}^s dp$$\n",
    "\n",
    "Here P is the space of all perturbations and $\\psi(P)$\n",
    "is the probability of a perturbation p. If all perturbations\n",
    "are equally likely this should be unity. $D_{a,p}^s$ is an\n",
    "evaluation function which gives the viability of the system\n",
    "under a perturbation. This is defined as\n",
    "\n",
    "$ D_{a,p}^s = \\left\\{\n",
    "  \\begin{array}{lr}\n",
    "    0 & p \\in A \\subset P\\\\\n",
    "    \\frac{f_a(p)}{f_a(0)} & p \\in P - A\n",
    "  \\end{array}\n",
    "\\right. $\n",
    "\n",
    "In the above definition the evaluation function is zero\n",
    "if p is in a set A which failed to maintain functions and\n",
    "it is given by a viability function when it is able to\n",
    "maintain function. The choice of the viability function\n",
    "is somewhat empirical and system dependent.\n",
    "\n",
    "Using this definition we can compare two different systems\n",
    "S1 and S2 and define one S1 as more robust than the S2 when we have\n",
    "\n",
    "$$R_{ap}^S1 > R_{ap}^S2$$\n",
    "\n",
    "The above comparison holds for a small perturbation space.\n",
    "When we consider a broad enough perturbation space then\n",
    "there will be trade-offs between robustness and fragility\n",
    "and the net contribution to the integral will vanish.\n",
    "\n",
    "$$\\Delta R_{aP}^{S1,S2} = \\int_{P}\\psi(P)(D_{a,p}^s1 - D_{a,p}^s2) dp = 0$$\n",
    "\n",
    "which means that $R_{aP}^S1 = R_{aP}^S1$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5><a id=\"detail_robust_example\">Example application of the general framework</a></h5>\n",
    "\n",
    "This definition of robustness was directly applied [(Kwon, 2008)](#kwon2008quantitative) to\n",
    "Boolean or piecewise-linear differential equation systems\n",
    "expressed as a directed graph G represented as expression\n",
    "and weights $G = (V, A)$. The vector V consists of elements\n",
    "$v_i$ which is a continuous variable taking values in the\n",
    "set $[-1,1]$ and A consists of set of ordered pairs with\n",
    "specific interaction weights $w_{ij}$. Under this definition\n",
    "the state of the node at time t+1 with values at t being\n",
    "$v_1(t),v_2(t),...,v_N(t)$ is given below:\n",
    "\n",
    "$$v_i(t+1) = f(\\sum_{j=1}^N w_{ij}v_j(t))$$\n",
    "\n",
    "To quantify the difference between two network states one\n",
    "can define a distance metric like $d(V_i,V_j) = \\sum_{i=1}^N (v_i - v_j)^2$.\n",
    "If the network is discrete and the variables $v_i(t)$ can\n",
    "only take values 0 or 1 then we can also use a popular\n",
    "metric, the Hamming distance which counts the number\n",
    "of changes needed to transform between the two networks.\n",
    "\n",
    "In this setting one can measure robustness by first\n",
    "generating a random initial state and then by making M\n",
    "set of perturbations on the initial values of $v_i$ and then\n",
    "finding how many of these cases keep **Hamming\n",
    "distance or the distance metric stay within a certain\n",
    "range**.\n",
    "\n",
    "We can similarly define fragility of the network to\n",
    "**point mutations** and **knockout mutations** by changing the\n",
    "intial state and the interaction weights and recalculating\n",
    "the distance metric. Again fragility of the network with\n",
    "respect to these mutations can be defined in terms of the\n",
    "fraction of the changes which keeps the distance metric\n",
    "within a specified limit.\n",
    "\n",
    "The key findings of this work are that:\n",
    "\n",
    "- The presence of feedback loops in the network has a\n",
    "  profound influence on its robustness. In particular\n",
    "  networks which were more robust against perturbations\n",
    "  were found to contain more positive feedback loops and\n",
    "  less negative feedback loops\n",
    "- The other finding verified the idea behind the robustness\n",
    "  fragility trade-offs by showing that networks are robust\n",
    "  when the nodes which are subjected to perturbations\n",
    "  involve a smaller number of feedback loops compared to\n",
    "  the nodes which contain no feedback."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<h5><a id=\"detail_robust_rel_evolve\">Evolvability</a></h5>\n",
    "\n",
    "Wagner [(Wagner, 2008)](#wagner2008robustness) defines the concept of neutral spaces or\n",
    "neutral networks which are metagraphs of biochemical\n",
    "networks (each node of a metagraph is a biochemical network\n",
    "with different topology or node states). In this context\n",
    "it becomes easy to define the concepts of robustness and\n",
    "evolvability. In particular Wagner\n",
    "distinguishes robustness of genotypes from that of phenotypes.\n",
    "In the context of neutral spaces, robustness to genotype\n",
    "is simply the number of 1-neighbors of a particular\n",
    "node in the metagraph that results in the same phenotype\n",
    "Whereas phenotype robustness is an average of all the\n",
    "1-neighbor of various nodes in the network that results\n",
    "in same phenotype. Evolvability is similarly defined\n",
    "as number of 1-neighbors which results in different\n",
    "phenotype. In particular this resolves the paradox in\n",
    "biology that says that increase in the robustness of\n",
    "genotypes results in a corresponding increase in phenotypic\n",
    "evolvability whereas intuition suggests it should be\n",
    "opposite.\n",
    "\n",
    "Using the idea of neutral spaces a mathematical framework\n",
    "is developed for discrete transcriptional networks. For a\n",
    "network of N nodes we can define the a network by the\n",
    "expression patterns of the genes $\\bar{S}(t) = (S_1(t), S_2(t),...,S_N(t))$\n",
    "The various interactions between the nodes is represented\n",
    "by a matrix $w_{ij}$ where each element represents the\n",
    "nature of interaction between nodes i and j. In general\n",
    "this matrix is not symmetric as the network is directed.\n",
    "The strength and the nature of interaction is indicated\n",
    "by the magnitude and sign of $w_{ij}$. Starting from an\n",
    "initial state $\\bar{S}(0)$ the system evolves to some\n",
    "equilibrium state $\\bar{S}_{\\inf}$. The initial state\n",
    "$\\bar{S}(0)$ is determined by conditions upstream of the\n",
    "network and enviromental conditions. The final state\n",
    "$\\bar{S}_{\\inf}$ will affect various downstream processes.\n",
    "Depending on a particular system any deviations from the\n",
    "state $\\bar{S}_{\\inf}$ can be deleterious."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": false
   },
   "source": [
    "<figure>\n",
    "<img src=\"Wagner_Robustness.png\" style=\"width: 300px;\"/>\n",
    "<figcaption>Network Interaction Matrix  [(Ciliberti, 2007)](#ciliberti2007robustness)</figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Within this framework Wagner first distinguishes between\n",
    "robustness to mutations with robustness to intrinsic\n",
    "system noise.\n",
    "Robustness to mutations is robustness with respect to\n",
    "changes in regulatory interaction like a change in topology\n",
    "or a change in interaction strength which is reflected\n",
    "by a different interaction matrix\n",
    "Robustness to noise is robustness with respect to gene \n",
    "expression patterns.\n",
    "Then he defines various measures corresponding\n",
    "to the type of robustness:\n",
    "\n",
    "**Robustness to mutations**: A measure of robustness to\n",
    "mutations $R_{\\mu}$ is the fraction of 1-neighbors in\n",
    "the metagraph that do not result in a change of $S_{\\inf}$.\n",
    "\n",
    "**Robustness to noise**: There are various measures of\n",
    "robustness with respect to noise which are defined. These\n",
    "are:\n",
    "**$R_{v1}$**: This is the fraction of single changes in\n",
    "the initial expression pattern $\\bar{S}(0)$ which preserves\n",
    "the equilibrium state.\n",
    "**$R_{v*}$**: This is the fraction of genes whose expression\n",
    "needs to change to decrease the probability of maintaining\n",
    "the final equilibrium state less than one-half.\n",
    "\n",
    "The key findings of this work are:\n",
    "\n",
    "- **Robustness to noise is strongly correlated with robustness to mutations** \n",
    "  so that one can be used instead of the other\n",
    "- Networks of varying regulatory strengths show a wide\n",
    "  distribution of robustness\n",
    "- Networks can gradually evolve to a state of high robustness\n",
    "  starting from one with low robustness"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>References:</h3> [top](#toc)\n",
    "<ol type=\"1\">\n",
    "<li><a id=\"battogtokh2002ensemble\">Battogtokh, D and Asch, DK and Case, ME and Arnold, J and Sch{\\\"u}ttler, H-B,(2002),An ensemble method for identifying regulatory circuits with special reference to the qa gene cluster of Neurospora crassa,Proceedings of the National Academy of Sciences,99,(26),16904--16909</a></li>\n",
    "<li><a id=\"daniels2008sloppiness\">Daniels, Bryan C and Chen, Yan-Jiun and Sethna, James P and Gutenkunst, Ryan N and Myers, Christopher R,(2008),Sloppiness, robustness, and evolvability in systems biology,Current opinion in biotechnology,19,(4),389--395</a></li>\n",
    "<li><a id=\"gutenkunst2007universally\">Gutenkunst, Ryan N and Waterfall, Joshua J and Casey, Fergal P and Brown, Kevin S and Myers, Christopher R and Sethna, James P,(2007),Universally sloppy parameter sensitivities in systems biology models,PLoS Comput Biol,3,(10),e189</a></li>\n",
    "<li><a id=\"kitano2007towards\">Kitano, Hiroaki,(2007),Towards a theory of biological robustness,Molecular systems biology,3,(1),137</a></li>\n",
    "<li><a id=\"kwon2008quantitative\">Kwon, Yung-Keun and Cho, Kwang-Hyun,(2008),Quantitative analysis of robustness and fragility in biological networks based on feedback dynamics,Bioinformatics,24,(7),987--994</a></li>\n",
    "<li><a id=\"ciliberti2007robustness\">Ciliberti, Stefano and Martin, Olivier C and Wagner, Andreas,(2007),Robustness can evolve gradually in complex regulatory gene networks with varying topology,PLoS Comput Biol,3,(2),e15</a></li>\n",
    "<li><a id=\"masel2010robustness\">Masel, Joanna and Trotter, Meredith V,(2010),Robustness and evolvability,Trends in Genetics,26,(9),406--414</a></li>\n",
    "<li><a id=\"pigliucci2008evolvability\">Pigliucci, Massimo,(2008),Is evolvability evolvable?,Nature Reviews Genetics,9,(1),75--82</a></li>\n",
    "<li><a id=\"wagner2008robustness\">Wagner, Andreas,(2008),Robustness and evolvability: a paradox resolved,Proceedings of the Royal Society of London B: Biological Sciences,275,(1630),91--100</a></li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
